src/pal/inc/unixasmmacros.inc

```diff
@@ -352,3 +353,159 @@ C_FUNC(\Name\()_End):
+        ldr \Reg, [sp, #\Offset]
+        .cfi_restore \Reg
+.endm
+
+.macro POP_CALLEE_SAVED_REGISTERS
+        pop {r4-r11, lr}
+        .cfi_adjust_cfa_offset -(4*9)
+        .cfi_restore r4
+        .cfi_restore r5
+        .cfi_restore r6
+        .cfi_restore r7
+        .cfi_restore r8
+        .cfi_restore r9
+        .cfi_restore r10
+        .cfi_restore r11
+        .cfi_restore lr
+.endm
+
+.macro PUSH_CALLEE_SAVED_REGISTERS
+        push {r4-r11, lr}
+        .cfi_adjust_cfa_offset (4*9)
+.endm
+
+.macro push_register Reg
+        push            {\Reg}
+        .cfi_adjust_cfa_offset 4
+.endm
+
+.macro push_argument_register Reg
+        push_register \Reg
+.endm
+
+.macro PUSH_ARGUMENT_REGISTERS
+        push {r0-r3}
+        .cfi_adjust_cfa_offset (4*4)
+.endm
+
+.macro pop_register Reg
+        pop            {\Reg}
+        .cfi_adjust_cfa_offset -4
+.endm
+
+.macro pop_argument_register Reg
+        pop_register \Reg
+.endm
+
+.macro POP_ARGUMENT_REGISTERS
+        pop {r0-r3}
+        .cfi_adjust_cfa_offset -(4*4)
+.endm
+
+.macro PROLOG_WITH_TRANSITION_BLOCK extraLocals = 0, saveFpArgs = 0, pushArgRegs = 0
+
+        __PWTB_FloatArgumentRegisters = \extraLocals
+        __PWTB_SaveFPArgs = \saveFpArgs
+
+        .if (__PWTB_SaveFPArgs == 1)
+                .if ((__PWTB_FloatArgumentRegisters % 8) != 0)
+                __PWTB_FloatArgumentRegisters = __PWTB_FloatArgumentRegisters + 4
+                .endif
+                
+                __PWTB_TransitionBlock = __PWTB_FloatArgumentRegisters + 8 * 8 + 8 // 8 floating point registers
+        .else                
+                .if ((__PWTB_FloatArgumentRegisters % 8) == 0)
+                __PWTB_FloatArgumentRegisters = __PWTB_FloatArgumentRegisters + 4
+                .endif
+                
+                __PWTB_TransitionBlock = __PWTB_FloatArgumentRegisters
+        .endif
+
+        __PWTB_StackAlloc = __PWTB_TransitionBlock
+
+        .ifc \pushArgRegs, DoNotPushArgRegs
+                PUSH_ARGUMENT_REGISTERS
+        .endif
+
+        PUSH_CALLEE_SAVED_REGISTERS
+        
+        alloc_stack     __PWTB_StackAlloc
+        
+        .if (__PWTB_SaveFPArgs == 1)
+                add r6, sp, #(__PWTB_FloatArgumentRegisters)
+                vstm r6, {d0-d7}
+        .endif
+        
+        //CHECK_STACK_ALIGNMENT
+        
+        END_PROLOGUE
+
+.endm
+
+.macro EPILOG_WITH_TRANSITION_BLOCK_RETURN
+
+        free_stack __PWTB_StackAlloc
+        POP_CALLEE_SAVED_REGISTERS
+        free_stack 16
+        bx lr
+                
+.endm
+
+.macro EPILOG_WITH_TRANSITION_BLOCK_TAILCALL
+
+        .if (__PWTB_SaveFPArgs == 1)
+                add r6, sp, #(__PWTB_FloatArgumentRegisters)
+                vldm r6, {d0-d7}
+        .endif
+        
+        free_stack __PWTB_StackAlloc
+        POP_CALLEE_SAVED_REGISTERS
+        POP_ARGUMENT_REGISTERS        
+        
+.endm
+#endif
+
+.macro SETALIAS New, Old
+        .equiv \New, \Old
+.endm
```

Comment:
We should probably refactor this into unixasmmacrosgeneric, amd64 and arm specific files.  It could get very #if-soup as is.